File: meteo.py
import os;
import requests;
import json;
from dateutil.parser import parse;

def parseKeys(keys, obj, txts):
	info = ""
	raw  = []
	for i, key in enumerate(keys):
		try:
			value = obj[key];
			info += txts[i].replace("VALUE", str(value))
			raw.append(value)
		except KeyError:
			raw.append(0)
			continue
	return (info, raw)

'''
METEO-PREDICCIONES
'''

headers = {
	"api_key": os.environ['aemet'],
	"Accept": "application/json"
}

'''
PARA SACAR LAS ESTACIONES DE MADRID:

url = "https://opendata.aemet.es/opendata/api/valores/climatologicos/inventarioestaciones/todasestaciones/"

r = requests.get(url, headers=headers)
body = json.loads(r.text)
url_data = body["datos"]

r = requests.get(url_data);
estaciones = json.loads(r.text)

for e in estaciones:
	if(e["provincia"]=="MADRID" and "MADRID" in e["nombre"]):
		print(e['nombre'], e['indicativo'], e['indsinop'])

MAD_st = list(filter(lambda e: e["provincia"]=="MADRID" and "MADRID" in e["nombre"], estaciones))
ids = list(map(lambda e: e["indicativo"], MAD_st))
'''

estaciones = [
	{"nombre": "MADRID AEROPUERTO", "id": "3129"},
	{"nombre": "MADRID, CIUDAD UNIVERSITARIA", "id": "3194U"},
	{"nombre": "MADRID, CUATRO VIENTOS", "id": "3196"},
	{"nombre": "MADRID, RETIRO", "id": "3195"}
]

for e in estaciones:
	'''
	DATOS ULTIMAS 24h - por estacion:

	"idema": "Indicativo climatógico de la estación meteorológia automática",
    "lon": "Longitud de la estación meteorológica (grados)",
    "lat": "Latitud de la estación meteorológica (grados)",
    "alt": "Altitud de la estación en metros",
    "fint": "Fecha hora final del período de observación, se trata de datos del periodo de la hora anterior a la indicada por este campo (hora UTC)",
    "ubi": "Ubicación de la estación. Nombre de la estación",
    
	PRECIPITACION
    "prec": "Precipitación acumulada, medida por el pluviómetro, durante los 60 minutos anteriores a la hora indicada por el período de observación 'fint' (mm, equivalente a l/m2)",
    X"pacutp": "Precipitación acumulada, medida por el disdrómetro, durante los 60 minutos anteriores a la hora indicada por el período de observación 'fint' (mm, equivalente a l/m2)"
    X"pliqtp": "Precipitación líquida acumulada durante los 60 minutos anteriores a la hora indicada por el período de observación 'fint' (mm, equivalente a l/m2)"
    X"psolt": "Precipitación sólida acumulada durante los 60 minutos anteriores a la hora indicada por el período de observación 'fint' (mm, equivalente a l/m2)"

	VIENTO
    "vmax":  "Velocidad máxima del viento, valor máximo del viento mantenido 3 segundos y registrado en los 60 minutos anteriores a la hora indicada por el período de observación 'fint' (m/s)",
    X"vmaxu": "Velocidad máxima del viento (sensor ultrasónico), valor máximo del viento mantenido 3 segundos y registrado en los 60 minutos anteriores a la hora indicada por el período de observación 'fint' (m/s)"
    "vv": "Velocidad media del viento, media escalar de las muestras adquiridas cada 0,25 ó 1 segundo en el período de 10 minutos anterior al indicado por 'fint' (m/s)",
    
    "dv": "Dirección media del viento, en el período de 10 minutos anteriores a la fecha indicada por 'fint' (grados)",
    "dmax": "Dirección del viento máximo registrado en los 60 minutos anteriores a la hora indicada por 'fint' (grados)",
    X"dmaxu": "Dirección del viento máximo registrado en los 60 minutos anteriores a la hora indicada por 'fint' por el sensor ultrasónico (grados)",

    "stdvv": "Desviación estándar de las muestras adquiridas de velocidad del viento durante los 10 minutos anteriores a la fecha dada por 'fint' (m/s)",
    "stddv": "Desviación estándar de las muestras adquiridas de la dirección del viento durante los 10 minutos anteriores a la fecha dada por 'fint' (grados)",

	OTROS
    "pres": "Presión instantánea al nivel en el que se encuentra instalado el barómetro y correspondiente a la fecha dada por 'fint' (hPa)",
    "pres_nmar": "Valor de la presión reducido al nivel del mar para aquellas estaciones cuya altitud es igual o menor a 750 metros y correspondiente a la fecha indicada por 'fint' (hPa)",
    
    "hr": "Humedad relativa instantánea del aire correspondiente a la fecha dada por 'fint' (%)",
    "vis": "Visibilidad, promedio de la medida de la visibilidad correspondiente a los 10 minutos anteriores a la fecha dada por 'fint' (Km)",
    
    TEMPERATURA
    "ts": "Temperatura suelo, temperatura instantánea junto al suelo y correspondiente a los 10 minutos anteriores a la fecha dada por 'fint' (grados Celsius)",
    "ta": "Temperatura instantánea del aire correspondiente a la fecha dada por 'fint' (grados Celsius)",
    "tamin": "Temperatura mínima del aire, valor mínimo de los 60 valores instantáneos de 'ta' medidos en el período de 60 minutos anteriores a la hora indicada por el período de observación 'fint' (grados Celsius)",
    "tamax": "Temperatura máxima del aire, valor máximo de los 60 valores instantáneos de 'ta' medidos en el período de 60 minutos anteriores a la hora indicada por el período de observación 'fint' (grados Celsius)",
    "tpr": "Temperatura del punto de rocío calculado correspondiente a la fecha 'fint' (grados Celsius)",
    
    "inso": "Duración de la insolación durante los 60 minutos anteriores a la hora indicada por el período de observación 'fint' (horas)"

	'''

	url = 'https://opendata.aemet.es/opendata/api/observacion/convencional/datos/estacion/{}'.format(e["id"])
	r = requests.get(url, headers=headers)
	body = json.loads(r.text)

	if(body["estado"]==404):
		print("Err getting data from {} ({})".format(e["nombre"], body["estado"]))
		# print("Response: {}".format(body["descripcion"]))
		continue
	else:
		print("{} {}".format(e["nombre"], body["descripcion"]))
		url_data = body["datos"]

	r = requests.get(url_data);
	data = json.loads(r.text)

	# data: lista con 23 valores, 1 por hora

	for d in data:
		date = parse(d["fint"]).strftime("%d-%m %H:%M")

		# ubi, hora, temp
		txts = [
			# "En VALUE {}.".format(date),
			" temp: VALUE°C,",
			" presión: VALUE (hPa),",
			" humedad: VALUE (%),",
			" viento: VALUE (m/s),",
			" precipitación: VALUE (l/m2)"
		]
		keys = [
			# "ubi",
			"ta",
			"pres",
			"hr",
			"vv",
			"prec"
		]
		(info, raw) = parseKeys(keys, d, txts)

		print(date, info);
		# print(raw)
	print()

File: crawler/meteo.js
const https = require("https");
const fs = require("fs");

function getData(id, callback) {
    let options = {
        hostname: "opendata.aemet.es",
        path: `/opendata/api/observacion/convencional/datos/estacion/${id}`,
        method: "GET",
        headers: {
            "api_key":  process.env.aemet,
            "Accept": "application/json"
        }
    };

    let req = https.get(options, (res) => {
        let buffer = "";

        res.on("data", (d) => {
            buffer += d;
        }).on("end", () => {
            const metadata = JSON.parse(buffer);

            if(metadata.estado==200){
                req = https.get(metadata.datos, (res) => {
                    buffer = "";

                    res.on("data", (d) => {
                        buffer += d;
                    }).on("end", () => {
                        const body = JSON.parse(buffer);

                        //console.log(estaciones.filter(e => e.id==id)[0].nombre, metadata.descripcion)
                        callback(parseData(body))
                    });
                });
            }
            else{
                callback([])
                //console.log(`Err getting data from ${estaciones.filter(e => e.id==id)[0].nombre} (${metadata["estado"]})`);
            }
        });
    });

    req.on("error", (e) => {
        console.error(e);
    });
    req.end();
}

function parseData(data) {

    //const date = parse(d["fint"]).strftime("%d-%m %H:%M")

    const txts = [
        " temp: VALUE°C,",
        " presión: VALUE (hPa),",
        " humedad: VALUE (%),",
        " viento: VALUE (m/s),",
        " precipitación: VALUE (l/m2)"
    ]
    const keys = [
        //"ubi",
        "ta",
        "pres",
        "hr",
        "vv",
        "prec"
    ]

    let day = []

    for(d of data) {
        let {verbose, values} = parseKeys(keys, txts, d)
        day.push([d.fint].concat(values))

        // TODO: pasar de array con valores a dict usable por console.log ¿?
        // console.log(d.fint, values)
    }
    // console.log()
    return day;

}

function parseKeys(keys, txts, obj) {
    let verbose = "", values  = [];

    keys.map((key, i) => {
        value = obj[key] ? obj[key] : 0;
        verbose += txts[i].replace("VALUE", String(value))
        values.push(value)

    });
    return {verbose, values}
}

function processData(alldata, estaciones, callback) {
    let i = 0;
    let response = {}

    const keys = [
        //"ubi",
        "temp (°C)",
        "presión (hPa)",
        "humedad (%)",
        "viento (m/s)",
        "precipitación (l/m2)"
    ]

    for(obj of alldata){
        // foreach station
        let {id, data} = obj;
        if(data.length==0) continue;
        data = data
            .map(d => d.splice(1))
            .map(d => {
                i = 0;
                return d.reduce((obj, attr) => {
                    obj[keys[i]] = attr
                    i++;
                    return obj
                }, {})
            });

        // console.log(estaciones.filter(e => e.id==id)[0].nombre, ":\n", data, '\n')
        estaciones = locations.filter(l => l[6]==id)
        estaciones.map(e => response[e[0]] = data[data.length-1]);
    }
    callback(response)
}

function NN(estaciones) {
    for (let zona of locations) {
        // zona: 35;Pza. del Carmen;40.41920833333333;-3.7031722222222223;4474524.28;440345.85;3195;MADRID, RETIRO
        nearest = getNN(zona[2], zona[3], estaciones)
        console.log(`${zona.join(";")};${nearest.id};${nearest.nombre}`)
    }
}

function getNN(x, y, estaciones) {
    x = Number(x), y = Number(y)
    let data = estaciones
        .sort((a, b) => {
            let ay = a.lon, ax = a.lat, by = b.lon, bx = b.lat;
            let a_distance = Math.sqrt(Math.pow((ax-x), 2) + Math.pow((ay-y), 2));
            let b_distance = Math.sqrt(Math.pow((bx-x), 2) + Math.pow((by-y), 2));
            // console.log(a_distance, b_distance, x, y, a.lat, a.lon, b.lat, b.lon)
            if (a_distance < b_distance)
                return -1;
            if (a_distance > b_distance)
                return 1;
            return 0;
        });
    return data[0];
}

function meteo(callback) {
    const estaciones = [
        {
            "nombre": "MADRID AEROPUERTO",
            "id": "3129",
            "lat": 40.466667,
            "lon": -3.555556
        },{
            "nombre": "MADRID, CIUDAD UNIVERSITARIA",
            "id": "3194U",
            "lat": 40.451667,
            "lon": -3.724167
        },{
            "nombre": "MADRID, CUATRO VIENTOS",
            "id": "3196",
            "lat": 40.375556,
            "lon": -3.786111
        },{
            "nombre": "MADRID, RETIRO",
            "id": "3195",
            "lat": 40.411944,
            "lon": -3.678056
        }
    ];

    let responses = [];
    let completed_requests = 0;


    fs.readFile("./csv/coordinates.csv", "utf8", (err, data) => {
        if (err) throw err;
        console.log("Loading locations");
        data
            .split("\n")
            .map(d => {
                // d = d.slice(0, -1); //unnecessary removal of \r ?
                locations.push(d.split(";"));
            });
        
        for (let e of estaciones) {
            let {id} = e;

            getData(id, (data) => {
                responses.push({id, data});
                completed_requests++;

                if(completed_requests == estaciones.length){
                    processData(responses, estaciones, callback);
                }
            });
        }
    });
}

let locations = [];

// meteo();

exports.m = meteo;
